instructure/canvas-ios
# Add support for Siri Shortcuts

Hi there 😄 For my Hackweek project, I added support for a few Siri Shortcuts to the Canvas apps. These are the shortcuts I added:

* Check Course Grade - Allows a user to obtain their grade in a course
* Get Courses - Allows a user to obtain a list of courses they are
  enrolled in
* Send Message - Allows a user to send a message to other user(s) using Siri

[Here is my Hackweek video](https://drive.google.com/file/d/1jfRA62baM3UJzdBLFinGZpayHt1Ncjs7/view?usp=sharing), if you're interested in seeing (a brief demo of) how this works.

This is my first contribution to the iOS project, so I'm hopeful that I've done things correctly, but please let me know if there is anything I should improve or do differently. Please also let me know if you have any questions.

Thank you!

**Note**: This change sets some properties to `public` inside of a few API models, in order
to avoid the complexity of interacting with CoreData from within the app
extension (this is the same pattern as what was previously done when
widgets were implemented).

---

refs: LS-2681
affects: Student, Teacher, Parent
release note: Added supoort for interacting with Canvas using Shortcuts and Siri.

test plan:
- in each of the Teacher, Parent, Student apps:
  - log in to the app
  - ask Siri something similar to "send a message using <app>"
    - verify "canvas", "student", "teacher", "parent", "canvas parent"
      are all recognized as app names
  - when prompted, provide a partial name of another user
    - if there are multiple matches, expect to see a disambiguation
      prompt (expect user avatars to be displayed)
    - if there is one match, expect to be asked to confirm the name
  - when prompted, provide message contents, and send the message
  - after sending the message, confirm that tapping on the "chat bubble"
    in the Siri UI opens the newly sent message in the app (Student and
    Teacher apps only)
- in the Student and Teacher apps:
  - verify that the 'Get Courses' action works correctly in the Shortcuts
    app
    - confirm that the filters work, and the results returned by the
      action are valid
  - verify that the 'Check Course Grade' action works correctly in the
    Shortcuts app
    - confirm that selecting a course manually works (list loads, fuzzy
      search works)
    - confirm that providing a course as a value from the 'Get Courses'
      action works
    - confirm that the returned results are valid
